### PR TITLE
Push events to influxdb heapster sink

### PR DESCRIPTION
**Pull request by saad-ali**
This implements the second part of #118 (follow up to PR #192).

Related Kubernetes Issues:
* GoogleCloudPlatform/kubernetes/issues/4432
* GoogleCloudPlatform/kubernetes/issues/5638

Screenshot showing events in InfluxDB (pulled by heapster from Kubernetes API server and pushed by heapster to influxdb):
influxdbevents

### CODE REVIEW

**Comment by rjnagal** [comment]
ok to test

**Comment by vishh** [comment]
Discussed with @saad-ali offline about the schema. He is working on updating the PR to use a more user friendly schema.

**Comment by saad-ali** [comment]
PTAL. I refactored the decoder such that there are separate decoders for events and metrics--they create common `timeseries` objects that the drivers (gcm and influxdb) can consume.
And based on your feedback, events and metrics now share the common schema (I added "pod_name").

New screenshot showing events in InfluxDB (pulled by heapster from Kubernetes API server and pushed by heapster to influxdb):
influxdb_events02

**Comment by jimmidyson** [comment]
Can't we just store the marshalled event object as JSON rather than a string which is much harder to parse for display, etc?

**Comment by vishh** [comment]
@saad-ali: I was thinking of just saving the event message as value. But it looks like you are storing a lot more data that that. If we think we need to store the entire event object, then I agree with @jimmidyson on marshalling the entire object into json. Sorry about the confusion.

**Comment by saad-ali** [comment]
I think there's other information in there that is useful. I'll switch it to JSON.

**Comment by jimmidyson** [comment]
Or we could create columns from the event data instead?

**Comment by vishh** [comment]
@jimmidyson: Having columns derived from Event object is confusing. I feel that the event object wasn't conceived to be human parsable.

**Comment by saad-ali** [comment]
@jimmidyson The original iteration of this PR was deriving columns from event objects. But after syncing with @vishh we agreed that this was contradictory to heapster's goals of having a common schema for all metrics/events. Hence this iteration.

**Comment by jimmidyson** [comment]
Understand now about the common schema. Is this schema documented anywhere? Surely the storage schema might differ between sinks?

Storing marshalled as JSON should cover what anyone needs to do with it.

**Comment by saad-ali** [comment]
@jimmidyson The idea was to try and maintain a common schema to the extent possible, and throw any additional information in to the value. Sink will always have the option to drop data that is irrelevant to them. The schema is documented here: [docs/storage-schema.md](https://github.com/GoogleCloudPlatform/heapster/blob/master/docs/storage-schema.md)

**Comment by saad-ali** [comment]
PTAL.

New screenshot:
influxdbevents05

File: sinks/api/types.go [hunk 1]
```diff
@@ -115,17 +123,14 @@ type MetricDescriptor struct {
...
 type Point struct {
-	// The name of the metric. Must match an existing descriptor.
-	Name string
-
-	// The labels and values for the metric. The keys must match those in the descriptor.
+	// The label keys and values for the metric.
 	Labels map[string]string
 
-	// The start and end time for which this data is representative.
+	// The start and end time for which this data is representative and the precision of the timestamps.
```
Comment (vishh) [inline sinks/api/types.go]: Precision is not included here.

File: sinks/gcm/driver.go [hunk 2]
```diff
@@ -141,43 +151,52 @@ type metricWriteRequest struct {
...
-func (self *gcmSink) StoreTimeseries(input []sink_api.Timeseries) error {
+func (self *gcmSink) StoreTimeseries(inputTimeseriesLists []*[]sink_api.Timeseries) error {
 	// Ensure the metrics exist.
-	for _, entry := range input {
-		metric := entry.Point
-		// TODO: Remove this check if possible.
-		if _, ok := self.exportedMetrics[metric.Name]; !ok {
-			return fmt.Errorf("unable to push unknown metric %q", metric.Name)
+	for _, inputTimeseriesList := range inputTimeseriesLists {
+		if inputTimeseriesList != nil {
```
Comment (vishh) [inline sinks/gcm/driver.go]: nit: How about reversing the condition and continuing if its nil?

File: sinks/gcm/driver.go [hunk 3]
```diff
@@ -141,43 +151,52 @@ type metricWriteRequest struct {
...
-		for key, value := range metric.Labels {
-			labels[fullLabelName(key)] = value
-		}
-
-		// TODO(vmarmol): Validation and cleanup of data.
-		// TODO(vmarmol): Handle non-int64 data types. There is an issue with using omitempty since 0 is a valid value for us.
-		if _, ok := metric.Value.(int64); !ok {
-			return fmt.Errorf("non-int64 data not implemented. Seen for metric %q", metric.Name)
+	for _, inputTimeseriesList := range inputTimeseriesLists {
+		if inputTimeseriesList != nil {
```
Comment (vishh) [inline sinks/gcm/driver.go]: Ditto as above.

File: sinks/influxdb/driver.go [hunk 4]
```diff
@@ -37,72 +37,130 @@ type influxdbSink struct {
...
 
-func (self *influxdbSink) StoreTimeseries(timeseries []sink_api.Timeseries) error {
+func (sink *influxdbSink) storeTimeseriesWithPrecision(timeseriesLists []*[]sink_api.Timeseries, precision sink_api.TimePrecision) error {
 	dataPoints := []*influxdb.Series{}
-	for index := range timeseries {
-		dataPoints = append(dataPoints, self.metricToSeries(&timeseries[index]))
+	for _, timeseriesList := range timeseriesLists {
+		if timeseriesList != nil {
+			for _, timeseries := range *timeseriesList {
+				if timeseries.TimePrecision == precision {
```
Comment (vishh) [inline sinks/influxdb/driver.go]: Why do we need this check? Why can't we use the precision that is passed via `timeseries.TimePrecision`?

File: sinks/influxdb/driver.go [hunk 5]
```diff
@@ -37,72 +37,130 @@ type influxdbSink struct {
...
+		err := sink.client.WriteSeriesWithTimePrecision(dataPoints, getInfluxdbPrecision(precision))
+		if err != nil {
+			glog.Errorf("failed to write time series to influxDB - %s", err)
+			for _, dataPoint := range dataPoints {
+				glog.Errorf("  detail: failed to write time series (%s) to influxDB with columns (%s)", dataPoint.Name, dataPoint.Columns)
+			}
+			sink.recordWriteFailure()
+			return err
+		}
+		glog.V(1).Info("Succesfully wrote %q timeseries to influxDB", precision)
```
Comment (vishh) [inline sinks/influxdb/driver.go]: This log file is misleading. Why are we logging precision in place of number of timeseries?

File: sinks/api/types.go [hunk 6]
```diff
@@ -148,24 +153,35 @@ type SupportedStatMetric struct {
...
-	Point            *Point
-	MetricDescriptor *MetricDescriptor
+	// The name of the series.
+	SeriesName string
+
+	// Precision of points timestamps.
+	TimePrecision TimePrecision
+
+	// The labels used by this series.
+	LabelKeys []string
```
Comment (vishh) [inline sinks/api/types.go]: Is this for optimization? Why can't we use all the labels defined as part of the point?

File: sinks/influxdb/driver.go [hunk 7]
```diff
@@ -37,72 +37,130 @@ type influxdbSink struct {
...
+	return influxdb.Second
+}
+
+func (sink *influxdbSink) timeSeriesToInfluxdbSeries(timeseries *sink_api.Timeseries) *influxdb.Series {
+	seriesName := timeseries.SeriesName
+
+	columns := []string{}
+	columns = append(columns, "time")            // Column 0 header
+	columns = append(columns, "value")           // Column 1 header
+	columns = append(columns, "sequence_number") // Column 2 header
```
Comment (vishh) [inline sinks/influxdb/driver.go]: How about sharing setting up of these standard columns between the two methods here?

File: sinks/api/eventdecoder.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	if data.Events != nil && len(data.Events) > 0 {
+		points := []Point{}
+		for _, event := range data.Events {
+			value, err := getEventValue(&event)
+			if err != nil {
+				return result, err
+			}
+			points = append(points, Point{
+				Labels: getEventLabels(&event),
+				Start:  event.LastTimestamp.Time,
```
Comment (vishh) [inline sinks/api/eventdecoder.go]: Should we round the timestamp here to a millisecond?

File: sinks/influxdb/driver.go (hunk 4 above)
Comment (saad-ali) [inline sinks/influxdb/driver.go]: It is required for filtering out other precisions. Each Timeseries can specify it's own precision and the influxDB client expects all points in a WriteSeriesWithTimePrecision call to have the same precision. So we filter the timeseries, and issue one WriteSeriesWithTimePrecision call per precision type (if that type is present).

File: sinks/api/types.go (hunk 1 above)
Comment (saad-ali) [inline sinks/api/types.go]: Fixed.

File: sinks/gcm/driver.go (hunk 2 above)
Comment (saad-ali) [inline sinks/gcm/driver.go]: Changed.

File: sinks/gcm/driver.go (hunk 3 above)
Comment (saad-ali) [inline sinks/gcm/driver.go]: Changed.

File: sinks/influxdb/driver.go (hunk 5 above)
Comment (saad-ali) [inline sinks/influxdb/driver.go]: See above comment about precision, essentially `storeTimeseriesWithPrecision` is called for each possible precision type (second, milli, micro). It filters any timeseries that match that precision, and issues a `WriteSeriesWithTimePrecision` for those points with that precision. So this log message is essentially saying that a time series with a given precision was written.

File: sinks/api/metricdecoder.go [hunk 9]
```diff
@@ -121,24 +123,36 @@ func (self *defaultDecoder) getContainerMetrics(container *source_api.Container,
 				} else {
 					startTime = stat.Timestamp
 				}
-				points := supported.GetValue(&container.Spec, stat)
-				for _, point := range points {
+				seriesName := supported.Name
+				if supported.Units.String() != "" {
+					seriesName = fmt.Sprintf("%s_%s", seriesName, supported.Units.String())
```
Comment (vishh) [inline sinks/api/metricdecoder.go]: I don't think GCM will like diverging from the registered series names. What ever series name we used while registering the events is what we should use here.
If Influxdb decides to mangle the series name with additional metadata, then it must be local to Influxdb.

**Comment by vishh** [comment]
@saad-ali: The PR looks good. Just a few comments. Some of the comments are related to sink functionalities and unfortunately they are not documented properly :(

File: sinks/api/types.go (hunk 6 above)
Comment (saad-ali) [inline sinks/api/types.go]: Yes, it is an optimization. Each `timeseries` can have multiple points, and any given point may specify some subset of labels for the timeseries. So, instead of iterating through all points and their lables to figure out the complete set of label keys, we have `LabelKeys` defined at the top defining all the labels that points in this `timeseries` may have (if any individual point defines a label that is not part of `labelKeys`, it is ignored, if any individual point does not provide a value for a label that is part of `labelKeys`, it is left empty).

File: sinks/influxdb/driver.go (hunk 5 above)
Comment (vishh) [inline sinks/influxdb/driver.go]: Hmm. Maybe rephrase to `Successfully wrote timeseries to InfluxDB with precision: %q`?

File: sinks/influxdb/driver.go (hunk 7 above)
Comment (saad-ali) [inline sinks/influxdb/driver.go]: Done.

File: sinks/influxdb/driver.go (hunk 4 above)
Comment (vishh) [inline sinks/influxdb/driver.go]: Ah! Why do we care about precision at an API level? Why not pass timestamps with max precision and round to a sane precision in the sinks?

File: sinks/gcm/driver.go (hunk 3 above)
Comment (vishh) [inline sinks/gcm/driver.go]: I found the same pattern in a few other files. Can you make the same change in other files as well?

File: sinks/api/eventdecoder.go (hunk 8 above)
Comment (saad-ali) [inline sinks/api/eventdecoder.go]: Done.

File: sinks/api/metricdecoder.go [hunk 10]
```diff
@@ -121,24 +123,36 @@ func (self *defaultDecoder) getContainerMetrics(container *source_api.Container,
...
+				if supported.Type.String() != "" {
+					seriesName = fmt.Sprintf("%s_%s", seriesName, supported.Type.String())
+				}
+				internalPoints := supported.GetValue(&container.Spec, stat)
+				points := []Point{}
+				for _, internalPoint := range internalPoints {
 					labels := copyLabels(labels)
-					for name, value := range point.labels {
+					for name, value := range internalPoint.labels {
 						labels[name] = value
```
Comment (vishh) [inline sinks/api/metricdecoder.go]: These labels might not show up as part of LabelKeys. Why not include all metric labels by default as part of labelkeys?

**Comment by saad-ali** [comment]
Closing this. #234 replaces this.
